hail/src/main/scala/is/hail/expr/ir/agg/ArrayElementRegionValueAggregator.scala

```diff
@@ -0,0 +1,269 @@
+package is.hail.expr.ir.agg
+
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.asm4s._
+import is.hail.expr.ir._
+import is.hail.expr.types.physical._
+import is.hail.io.{CodecSpec, InputBuffer, OutputBuffer}
+import is.hail.utils._
+
+import is.hail.asm4s.coerce
+
+// initOp args: initOps for nestedAgg, length if knownLength = true
+// seqOp args: array, other non-elt args for nestedAgg
+
+object ArrayElementState {
+  def create(mb: EmitMethodBuilder, aggs: Array[StagedRegionValueAggregator], knownLength: Boolean): ArrayElementState =
+    ArrayElementState(aggs.map(_.createState(mb)), mb.newField[Region], mb.newField[Long], knownLength)
+}
+
+case class ArrayElementState(nested: Array[RVAState], r: ClassFieldRef[Region], off: ClassFieldRef[Long], knownLength: Boolean) extends RVAState {
+  private val container: StateContainer = StateContainer(nested, r)
+  private val arrayType: PArray = PArray(container.typ)
+
+  private val nStates: Int = nested.length
+  val mb: EmitMethodBuilder = nested.head.mb
+
+  val typ: PTuple = PTuple(FastIndexedSeq(container.typ, arrayType))
+
+  val lenRef: ClassFieldRef[Int] = mb.newField[Int]
+  val idx: ClassFieldRef[Int] = mb.newField[Int]
+
+  val srvb = new StagedRegionValueBuilder(er, typ)
+
+  private def regionOffset(eltIdx: Code[Int]): Code[Int] = (eltIdx + 1) * nStates
+  private def regionFromIdx(eltIdx: Code[Int], stateIdx: Int): Code[Region] = container.getRegion(regionOffset(eltIdx), stateIdx)
+
+  private val initStatesOffset = typ.loadField(region, off, 0)
+  private def initStateOffset(idx: Int): Code[Long] = container.getStateOffset(initStatesOffset, idx)
+  private def initStateAddress(idx: Int): Code[Long] = container.loadStateAddress(initStatesOffset, idx)
+
+  private def statesOffset(eltIdx: Code[Int]): Code[Long] = arrayType.loadElement(region, typ.loadField(region, off, 1), eltIdx)
+  private def stateAddressOffset(eltIdx: Code[Int], stateIdx: Int): Code[Long] = container.getStateOffset(statesOffset(eltIdx), stateIdx)
+  private def eltState(eltIdx: Code[Int], stateIdx: Int): Code[Long] = container.loadStateAddress(statesOffset(eltIdx), stateIdx)
+
+  def initLength(len: Code[Int]): Code[Unit] = {
+    val srvb2 = new StagedRegionValueBuilder(er, arrayType)
+    Code(
+      lenRef := len,
+      region.setNumParents((lenRef + 1) * nStates),
+      srvb2.start(lenRef),
+      Code.whileLoop(srvb2.arrayIdx < lenRef,
+        container.scoped(regionOffset(srvb2.arrayIdx)){ (i, s) =>
+          s.copyFrom(initStateAddress(i)) },
+        container.addState(srvb2),
+        srvb2.advance()),
+      typ.setFieldPresent(region, off, 1),
```

Comment:
I'm missing where the init states get initialized. Shouldn't there be a `typ.setFieldPresent(region, off, 0)` somewhere?

Reply:
There's basically two steps to the state initialization---
1. The init states get initialized with their respective init ops. The array of agg states is set to missing, since we don't necessarily know the length of the array.
2. We acquire the length of the array, and copy the init states into an array of that length. That's why the `setFieldPresent` is there---we expect the entire tuple to be present and initialized, but the first field to be missing, and we need to toggle that.

This is set up because the ArrayAggElements node provides us an optional `knownLength` argument---if it exists, then we'll initialize the length in the init op, but if it doesn't, we need to pick up the length from the first array that we see during the seqOp.

Reply:
Okay, I think I see. So the init states are never referenced after the second initialization step? Where does the first step happen?

Reply:
The first step happens in the initOp.